src/dict_manager.rs

```diff
@@ -0,0 +1,527 @@
+use crate::{
+    ids::PyTypedId,
+    memory_segments::PySegmentManager,
+    relocatable::{PyMaybeRelocatable, PyRelocatable},
+    utils::to_py_error,
+};
+use cairo_rs::{
+    hint_processor::builtin_hint_processor::dict_manager::DictManager,
+    types::relocatable::Relocatable,
+};
+use num_bigint::BigInt;
+use pyo3::{exceptions::PyKeyError, prelude::*};
+
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
+
+#[pyclass(unsendable)]
+pub struct PyDictManager {
+    manager: Rc<RefCell<DictManager>>,
+}
+
+#[pyclass(unsendable)]
+pub struct PyDictTracker {
+    manager: Rc<RefCell<DictManager>>,
+    key: Relocatable,
+}
+
+impl Default for PyDictManager {
+    fn default() -> Self {
+        PyDictManager::new()
+    }
+}
+
+#[pymethods]
+impl PyDictManager {
+    #[new]
+    pub fn new() -> Self {
+        PyDictManager {
+            manager: Rc::new(RefCell::new(DictManager::new())),
+        }
+    }
+
+    pub fn new_dict(
+        &self,
+        segments: &mut PySegmentManager,
+        initial_dict: HashMap<BigInt, BigInt>,
+        py: Python,
+    ) -> PyResult<PyObject> {
+        Ok(PyMaybeRelocatable::from(
+            self.manager
+                .borrow_mut()
+                .new_dict(&mut segments.vm.borrow_mut(), initial_dict)
+                .map_err(to_py_error)?,
+        )
+        .to_object(py))
+    }
+
+    pub fn new_default_dict(
+        &mut self,
+        segments: &mut PySegmentManager,
+        default_value: BigInt,
+        initial_dict: Option<HashMap<BigInt, BigInt>>,
+        py: Python,
+    ) -> PyResult<PyObject> {
+        Ok(PyMaybeRelocatable::from(
+            self.manager
+                .borrow_mut()
+                .new_default_dict(&mut segments.vm.borrow_mut(), &default_value, initial_dict)
+                .map_err(to_py_error)?,
+        )
+        .to_object(py))
+    }
+
+    pub fn get_tracker(&mut self, dict_ptr: &PyTypedId) -> PyResult<PyDictTracker> {
+        let ptr_addr = dict_ptr.hint_value.clone();
+        self.manager
+            .borrow()
+            .get_tracker(&ptr_addr)
+            .map_err(to_py_error)?;
+        Ok(PyDictTracker {
+            manager: self.manager.clone(),
+            key: ptr_addr,
+        })
+    }
+}
+
+#[pymethods]
+impl PyDictTracker {
+    #[getter]
+    pub fn get_current_ptr(&self, py: Python) -> PyResult<PyObject> {
+        Ok(PyRelocatable::from(
+            self.manager
+                .borrow_mut()
+                .get_tracker_mut(&self.key)
+                .map_err(to_py_error)?
+                .current_ptr
+                .clone(),
+        )
+        .into_py(py))
+    }
+
+    #[getter]
+    pub fn get_data(&self, py: Python) -> PyObject {
+        PyDictTracker {
+            manager: self.manager.clone(),
+            key: self.key.clone(),
+        }
+        .into_py(py)
+    }
```

Comment:
This doesn't look like it's getting any kind of `data` member. Instead, this is effectively a Python object copy.

Reply:
I'm new to the RC structure, but I believe with the manager being an Rc<...>, when I'm cloning the data I'm actually taking the same pointer and increasing the reference counter by one. Am I understanding this correctly? 
If the above is true, wouldn't `get_data` output an object referencing the same tracker in the input , also being able to modify the data inside the original input?

Reply:
1. Yes, `Rc<...>` returns a reference to the same underlying object;
2. That's true for the field, but it's still creating a new Python object (and in particular it lives on the heap);
3. What I'm asking about is more about the type of the return. It's OK that the implementation simply refers to the same `cairo-rs` object, but it doesn't make a lot of sense that `get_data` returns the same Python type I think. I mean, think about the API usage, why would you call `get_data` to get the same `PyDictTracker` you're calling the method on?